### ISSUE TITLE
feat: add cache prefix

### ISSUE THREAD

**Issue by ernado** [actions/setup-go#358]
**Description:**
Add cache key prefix, like that:
```yaml
- name: Install Go
  uses: actions/setup-go@v4
  with:
    go-version: '1.20'
    cache: true
    cache-prefix: 'test-${{ matrix.arch }}-${{ matrix.flags }}-'
```

**Justification:**
Cache from multiple workflows can be mixed.
For example, there can be a workflow that only downloads dependencies.
This will produce cache without `.cache/go-build`, which will be used by
another workflows, effectively disabling build cache.

This was the cause in #357

**Are you willing to submit a PR?**

I don't know typescript well, but I'm learning.

**Workaround**

Use `actions/cache@v3` manually, something like that:

```yaml
name: manual

on:
  push:
    branches: [main]

jobs:
  build:
    runs-on: ubuntu-latest
    env:
      cache_name: manual

    steps:
      - uses: actions/checkout@v3

      - uses: actions/setup-go@v4
        with:
          go-version: "1.20.x"
          cache: false

      - name: Get Go environment
        run: |
          echo "cache=$(go env GOCACHE)" >> $GITHUB_ENV
          echo "modcache=$(go env GOMODCACHE)" >> $GITHUB_ENV

      - name: Set up cache
        uses: actions/cache@v3
        with:
          path: |
            ${{ env.cache }}
            ${{ env.modcache }}
          key: ${{ env.cache_name }}-${{ runner.os }}-go-${{ hashFiles('**/go.sum') }}-${{ hashFiles('**/go.mod') }}
          restore-keys: |
            ${{ env.cache_name }}-${{ runner.os }}-go-

      - run: go env GOCACHE
      - run: go build -v
```

**Cons**
Cache configuration options can possibly lead to feature bloat, so feel free to decline this feature proposal.

**Comment by marko-zivic-93** [comment]
Hello @ernado 
Thank you for sharing this feature request with us. We will take a look at it, and come back to you as soon as we have some info 🙂

**Comment by erezrokah** [comment]
Hi 👋 Are you open for a contribution for this issue? We're experiencing it as well and using the suggested workaround (using `actions/cache`).

**Comment by erezrokah** [comment]
Opened https://github.com/actions/setup-go/pull/366, feedback appreciated

**Comment by lzap** [comment]
This could help to workaround the [problem we have on our project](https://github.com/actions/setup-go/issues/316#issuecomment-1551238935).

**Comment by andig** [comment]
I've recently noticed a fair bit of slow-down in our parallel builds from 2:30 to 3:30, see https://github.com/evcc-io/evcc/actions/runs/5124962244 for an example. It shows that both build and test steps- despite stable dependencies- will download a fair bit of packages that should already be there:

```
fatal: No names found, cannot describe anything.
Version: 8acee30 8acee30 2023-05-30_18:24:01
CGO_ENABLED=0 go build -v -tags=release -ldflags='-X github.com/evcc-io/evcc/server.Version=8acee30 -X github.com/evcc-io/evcc/server.Commit=8acee30 -s -w'
go: downloading github.com/Masterminds/sprig/v3 v3.2.3
go: downloading github.com/andig/gosunspec v0.0.0-20211108155140-af2e73b86e71
go: downloading github.com/bogosj/tesla v1.1.1-0.20230430222423-0d6e63ee90c9
...
```

It seems likely that this might be a conflict with the parallel `clean` job that just installs the tools and does some basic tests:

```
Run make install
fatal: No names found, cannot describe anything.
go install $(go list -f '{{join .Imports " "}}' tools.go)
go: downloading github.com/golang/mock v1.6.0
go: downloading github.com/dmarkham/enumer v1.5.8
go: downloading github.com/pascaldekloe/name v1.0.1
go: downloading golang.org/x/tools v0.9.1
go: downloading golang.org/x/mod v0.10.0
go: downloading golang.org/x/sys v0.8.0
```

It seems that the situation described in this issue might be quite common.

**Comment by andig** [comment]
And here is the result applying this issue's PR: https://github.com/evcc-io/evcc/actions/runs/5126503582. Build time reduced from 3min to 1min at the cost of 3 separate caches. Worth the efforts.

Any chance to get this reviewed/merged?

**Comment by dmitry-shibanov** [comment]
Hello everyone. We would not like to implement such kind of logic, because the initial caching logic for setup-go should cover the most often use cases. If you need more flexibility with cache, we would like to recommend [actions/cache](https://github.com/actions/cache). It's relevant to setup-go [ADR](https://github.com/actions/setup-go/blob/main/docs/adrs/0000-caching-dependencies.md#goals--anti-goals).

For now I'm going to close the issue.

**Comment by erezrokah** [comment]
Hi @dmitry-shibanov and thanks for the comment. Are you open to adding docs for the cases the caching logic in setup-go doesn't cover? It does seem like a gotcha that should be documented (based on the backlinks to this issue) as the caching is now enabled by default.

Something in the lines of:
> Please note that if you use `setup-go` from multiple jobs (e.g. `build` and `test`) you'll need to disable caching and use `actions/cache`, otherwise caching will not work as expected

**Comment by ernado** [comment]
Agree with @erezrokah, adding this to docs (and probably providing a reference workaround) would be very helpful.

Also agree with @dmitry-shibanov that this feature should not be implemented, because it is hard to cover all use-cases, so just let user to handle caches manually (like we did in our projects).

**Comment by erezrokah** [comment]
> Plus this needs to cover which folders to cache for Go best practices.

And also explain to use the Go version from `setup-go` outputs in the cache key

**Comment by andig** [comment]
@erezrokah would you potentially consider rebasing your PR and keeping your fork going forward?

**Comment by erezrokah** [comment]
> @erezrokah would you potentially consider rebasing your PR and keeping your fork going forward?

I'm not sure that would the best solution. If there are docs that users can copy paste the get the same functionality then my PR would be mostly nice to have I think, and won't justify maintaining a fork

**Comment by andig** [comment]
@dsame thank you for reopening as investigation. Is there anything we could contribute?

**Comment by lucacome** [comment]
I removed `actions/cache` from my workflows when the caching was first implemented and started noticing that it didn't seem to use the cache anymore https://github.com/actions/setup-go/issues/130#issuecomment-1207103093

> We don't pursue the goal to provide wide customization of caching in scope of actions/setup-go action. The purpose of this integration is covering ~90% of basic use-cases.

I'm not sure how you arrived at that "~90% of basic use-cases" but it seems unlikely to me that 90% of projects (if that's what it means) don't have concurrent jobs running in a pipeline. And since caching now is enabled by default it means it won't work by default. I don't think we're asking for fine-tuning of the caching mechanism here, but just for a way to make the caching work in even more cases.

Whatever that 10% is, still doesn't seem an irrelevant number especially when it can be fixed by a 2-line PR

**Comment by dsame** [comment]
Hello @andig, @ernado 

The issue relates to the feature request for optional build directory caching https://github.com/orgs/actions/projects/19/views/1?filterQuery=assignee%3Adsame&pane=issue&itemId=35194372 
and should be resolved with it. 

The advantage of skipping build directory at all seems to be more preferable than having key-prefix because it allows to have multiple build steps in the one job instead of multiple jobs installing the same go version just for sake of getting different caches having the almost same content due to the common dependencies.

Although there might be a case where caching build directory brings significant improving over skipping it. At the moment there's no any discover, but it would be useful know different opinions to make a final decision

Adding the cache prefix brings significant drawback of using log of storage with multiple caches and hitting the limit very soon, making the caching impossible at all. 

Thanks in advance for sharing your thoughts.

**Comment by andig** [comment]
Seems I can't follow the link. Would like to read up on build directory caching to understand the proposal better.

**Comment by candiduslynx** [comment]
@dsame 

> Although there might be a case where caching build directory brings significant improving over skipping it. At the moment there's no any discovered, but it would be useful know different opinions to make a final decision

We want to be able to cache test results as in some cases clean testing would run for ~10 minutes, whilst the cached results will be available in seconds.

**Comment by andig** [comment]
I'm still failing to understand the proposal of "optional build directory caching". The link in https://github.com/actions/setup-go/issues/358#issuecomment-1674545346 does not work.

> We want to be able to cache test results as in some cases clean testing would run for ~10 minutes

We're also caching:
- downloaded Go modules
- linter results which is otherwise very slow

We also have different build steps like generating assets, updating docs and checking repo is in "clean" state that are all parallelised. Some of these steps use Go-ased tools but would not necessarily download modules or fully populate the build cache. Using any of these steps for heavier tasks like build or test has ended up with cache hits on half-populated caches.

**Comment by erezrokah** [comment]
> multiple build steps in the one job instead of multiple jobs

Thanks for sharing @dsame ❤️ Can you maybe share a before/after workflow example with the optional build step so we can understand the proposal better?

Won't that mean that the different steps run in serial as opposed to parallel build jobs? A cache prefix tries to solve using `setup-go` in parallel build jobs with caching enabled. I'm not sure how putting all steps in a single job can allow running those in parallel.

Please note that at least in our case we also have completely different workflows (not only jobs) that run in parallel, and each workflow needs a distinct cache (for example, one used for testing invoked on pull request triggers, and other used for releasing artifacts invoked on tag push triggers).

 > Adding the cache prefix brings significant drawback of using log of storage with multiple caches and hitting the limit very soon, making the caching impossible at all.
 
In our case we treat managing cache limits as a distinct problem to solve (especially as we have a monorepo with many components, each one with a different cache), unrelated to the requirement to have separate caches for jobs.

There are official guides (see https://docs.github.com/en/actions/using-workflows/caching-dependencies-to-speed-up-workflows#force-deleting-cache-entries) on how to manage cache limits, and it might be better to leave the decision on how to manage cache limits to users.

Regardless of cache limits and this issue, I would prefer to have a cache miss than a cache hit on a cache created from an unrelated job, as at least a cache miss will create consistent results.

**Comment by dsame** [comment]
@andig the complex cases like the one you have described should use [@actions/actions-cache ](https://github.com/actions/cache), see the [samples](https://github.com/actions/cache/blob/main/examples.md#go---modules) as starting point

**Comment by andig** [comment]
@dsame after looking at the samples I still see this an ill-advised tradeoff. We require people to think about cache keys, restore keys, pathes, hashfiles etc instead of providing them the absolutely simple and straight-forward capability of separating their caches per prefix which is a single-line PR.

More so, with any change in actions/cache or the way how Go dependencies should be cached, we would require everyone to update their manual cache settings once more. The fact that even documenting the advanced cases has been ongoing for six months now indicates that this is not straight-forward to get right.

I can only suggest to rethink this decision. I do not object to adding additional docs for advanced use cases, but it would be super handy if usage for semi-simple cases could be simplified.

**Comment by andig** [comment]
Friendly ping @dsame would appreciate a comment if these concerns are valid?

**Comment by macie** [comment]
I have similar issue as @jsirianni - different branches share the same cache. When my feature branch starts using first external module, `go.sum` file was created and cache misses starts to appear on other branches (because they don't have external dependencies/`go.sum`).

From my (user of _setup-go_) perspective, cache integration covers only one case: trunk-based development with sequential (non-parallel) jobs. But it isn't clear from the beginning.

GitHub encourages working on feature branches. Even in this repo nearly all pull requests come from feature branches. So I share @lucacome opinion, that current defaults may be unsuitable for far more than 10% users of _setup-go_.

**Comment by relaxdiego** [comment]
For those that don't want to deal with the unnecessary complexity of `actions/cache`, you can use this as a workaround:

```yaml
  job_a:
    steps:
      - uses: actions/setup-go@v5
        with:
          cache-dependency-path: |
            "${{ env.WORKING_DIR }}/go.sum"
            .cache/buster-for-job-a

  job_b:
    steps:
      - uses: actions/setup-go@v5
        with:
          cache-dependency-path: |
            "${{ env.WORKING_DIR }}/go.sum"
            .cache/buster-for-job-b
```

Notice the additional files in `cache-dependency-path` named `.cache/buster-for-job-a` and `.cache/buster-for-job-b`. Make sure those files have totally different contents. They can contain anything as long as they're different. The effect of that is that each job's cache key will have a unique hash suffix.

Another "feature" that this workaround provides is that you can force a rebuild of the job's cache by modifying its respective "cache buster" file.

**Comment by andig** [comment]
Worth mentioning that https://github.com/actions/setup-go/pull/416 is still open almost 6 months later. Splitting caches by build remains a complex task.

**Comment by techfg** [comment]
Ran into this issue.  Recently split out CI workflow on a monorepo with go/javascript/typescript projects into multiple jobs (still a single workflow), one for `format & lint` and one for `build & test` and all was good.  However, after upgrading to go 1.23.5, workflow times increased significantly.  After considerable time troubleshooting, I noticed in our logs that the go cached item size had decreased from prior runs.  Stumbled upon this issue which explained the situation thankfully.  

In short, the fmt/lint job finishes first, but the modules used by `go fmt` are different than the modules used by `go build/test` which meant that the `build & test` job was downloading modules on every run (because `fmt/lint` finished first and created the cache item).  The reason this wasn't a problem prior to the bump to 1.23.5 was because even after the separation of the jobs was made, the cached item was created for 1.23.4 when everything was still one job, so the cached item had everything that build used/needed.

Very disappointing that docs on this action don't even mention the current behavior and gotcha's for multi-job scenarios - this would have saved me (and others from posts in this issue) a lot of time.

Regarding solutions, to summarize what's covered in this issue:

1. Add a cache-prefix option to `actions/setup-go`
2. Use actions@cache directly
3. `go mod download` every module in a project in every job that uses go (this would ensure that any job that uses go will have all modules in the cached item even though that job may/may not use it).
4. Ensure that `actions/setup-go` cache-dependency graph contains at least one "unique" file per "unique" cache required (e.g., per job in most cases I would expect)

> Hello everyone. We would not like to implement such kind of logic, because the initial caching logic for setup-go should cover the most often use cases. If you need more flexibility with cache, we would like to recommend [actions/cache](https://github.com/actions/cache). It's relevant to setup-go [ADR](https://github.com/actions/setup-go/blob/main/docs/adrs/0000-caching-dependencies.md#goals--anti-goals).
> 
> For now I'm going to close the issue.

@dmitry-shibanov - There has been a lot of activity on this issue since you posted the above.  @erezrokah submitted #366  (essentially a one-line change) to solve for these use cases and @dsame submitted #416 after the decision was made to not add a "cache-prefix" option.  As you can see from others throughout this issue and the use case I describe above, even a single workflow with multiple jobs runs into these caching issues that would be trivial to solve with a new `cache-prefix` option on the action which would not change any default behavior of the currently existing action.  

Can #366 be re-considered (if docs were added to it indicating the specific use cases for specifying a cache-prefix and the downsides/impact that come with it) or at least #416 reviewed and merged so that people using the action are aware of the behavior/issues and can be saved from hours of investigation/troubleshooting?  

I'm happy to contribute on either approach if I know the team is open to accepting one of them.

**Comment by peterebden** [comment]
I agree with the last message; I think this would be very useful. So much so that I think maybe there should be a default prefix including `${{ github.workflow }}` - it's a very common setup to have more than one workflow, commonly one for linting (probably golangci-lint) and one for running tests. We have more or less that (in an internal repo so I can't share here) where the lint workflow is consistently faster, so it tends to win the race to populate the cache - which it puts less into than the full test run does. Hence the cache ends up being less useful for the slower run where you need it more. We've had a lot of improvement by swapping to actions/cache, but it is also more work & more YAML than the two lines we had for setup-go previously.

It's nice that this action is aiming for simplicity, but I think that there's a very common use case here that's being under-served; it takes a bit of thinking to realise what's going on given that by default all the caching settings are handled for you. And that assumes that you even realise something is wrong, I suspect very many builds are just suboptimal in this way and their owners don't really realise.

**Comment by jsok** [comment]
>  As you can see from others throughout this issue and the use case I describe above, even a single workflow with multiple jobs runs into these caching issues that would be trivial to solve with a new cache-prefix option on the action which would not change any default behavior of the currently existing action.

> Can [#366](https://github.com/actions/setup-go/pull/366) be re-considered (if docs were added to it indicating the specific use cases for specifying a cache-prefix and the downsides/impact that come with it) or at least [#416](https://github.com/actions/setup-go/pull/416) reviewed and merged so that people using the action are aware of the behavior/issues and can be saved from hours of investigation/troubleshooting?

The sheer amount of effort and time gone into **not** merging #366 is baffling. The amount of time, words and energy which have gone into this issue outweighs _just supporting a cache prefix_ option by an order of magnitude.
@actions/setup-actions-team @HarithaVattikuti  please reconsider https://github.com/actions/setup-go/blob/main/docs/adrs/0000-caching-dependencies.md#goals--anti-goals to include this improvement.

**Comment by andig** [comment]
Fully agreed. I know I shouldn‘t be writing this and just hit +1, but this just isn‘t enough after all these years!

**Comment by erezrokah** [comment]
Author of https://github.com/actions/setup-go/pull/366 here 👋 It has been a minute since I opened that [PR](https://github.com/actions/setup-go/pull/366)

I'm happy to re-open it, but I think it's quite easy nowadays to ask your favorite AI tool to disable caching for all `setup-go` workflows and use `actions/cache` instead with a key per job.

The [docs](https://github.com/actions/setup-go/blob/main/docs/adrs/0000-caching-dependencies.md#goals--anti-goals) for `setup-go` do say 

> The purpose of this integration is covering ~90% of basic use-cases

I can look up `setup-go` usages in public repositories to see if that's the current status. I doubt it is since for caching to work as expected using `setup-go` current non customizable key cache you're basically limited to a single job in the entire repository that can use `setup-go` with caching enabled.

**Comment by andig** [comment]
It's a bit sad, but "let the ai work around it 4-year old deficiencies" is the new pattern.

**Comment by v-priya-kinthali** [comment]
Hello everyone👋,

Thank you all for the thoughtful proposal and the detailed discussion on this thread. After a thorough review and internal discussion, we have decided **not to move forward** with adding a cache prefix option to `setup-go`. Our reasoning is as follows:

The caching in `setup-go` is intentionally scoped to cover roughly the most common basic use cases. Adding configurable cache key options moves into territory that [`actions/cache`](https://github.com/actions/cache) is purpose-built to handle. Adding these would increase complexity, testing surface, and long-term maintenance burden.

This use case is already fully supported today through [`actions/cache`](https://github.com/actions/cache), which is designed specifically for advanced and customized cache management. The workaround examples provided in this thread are a great reference, and we kindly encourage users with these needs to follow that approach.

**Option 1 — Stay in `setup-go` (unique cache per job)**

Add a small per-job file to `cache-dependency-path`. Since `setup-go` hashes every listed file, this gives each job its own cache key:

```yaml
- uses: actions/setup-go@v7
  with:
    go-version: '1.22'
    cache-dependency-path: |
      go.mod
      .github/cache-${{ github.job }}
```

Create one file per job with any differing content (e.g. `.github/cache-build`, `.github/cache-test`).

**Option 2 — Use `actions/cache` directly**

Disable `setup-go` caching and put your prefix straight in the key:

```yaml
- uses: actions/setup-go@v7
  with:
    go-version: '1.22'
    cache: false

- uses: actions/cache@v6
  with:
    path: |
      ~/.cache/go-build
      ~/go/pkg/mod
    key: ${{ github.job }}-${{ runner.os }}-go-${{ hashFiles('**/go.mod') }}
    restore-keys: |
      ${{ github.job }}-${{ runner.os }}-go-
```

Please note that per-key caches increase storage usage and may reach the repository cache limit sooner, so choose the approach that best fits your needs.

We are grateful for your contribution and understanding, and we hope you continue sharing ideas with us. With that, we are proceeding with closing this issue as **not planned**. Thanks again!

**Comment by jsok** [comment]
@priya-kinthali respectfully:

> Adding these would increase complexity, testing surface, and long-term maintenance burden.

This is a bit of a stretch, https://github.com/actions/setup-go/pull/366/changes was `+32, -2` PR 😀 

> intentionally scoped to cover roughly the most common basic use cases

Given the existence of this issue, I think you should re-evaluate what the most common use cases are again since https://github.com/actions/setup-go/blob/main/docs/adrs/0000-caching-dependencies.md was authored ~4 years ago. I'd say your assessment is out of touch with what the community is arguing as a common use case.

**Comment by candiduslynx** [comment]
@philip-gai any chance of reconsideration or is this a finalised decision?

While I appreciate the workarounds presented, the
> small per-job file

looks like a hack at best, an afterthought dismissal at worst.

And while `actions/cache` indeed solves the issue, this is still the most common use for Go when you are building multiple binaries/packages at a time (monorepo pattern) not to be addressed directly.

**Comment by macie** [comment]
But isn't basic cache [still broken as mentioned earlier](https://github.com/actions/setup-go/issues/358#issuecomment-1827836567)? GH encourages [flow based on branches](https://docs.github.com/en/get-started/using-github/github-flow), which current caching strategy isn't aligned with.

If this behavior wasn't silently fixed in the meantime, it would be nice to clearly specify in docs what is covered by basic caching, as people using branches will waste their time on debugging cryptic errors. Or just remove caching support, to not mislead the users about this action capabilities

**Comment by andig** [comment]
And of course this is a well-understood issue. Even the Go Weekly contained an article: https://www.cloudx.ai/posts/setup-go.

But the Go action's team simply rather likes to discuss for ages rather than adding a 3-lines change that gives power to those who need it without breaking things for everybody else. So life is miserable for those in need with just a little more control.

This is so annoying...
